btc/utils.go

```diff
@@ -150,14 +183,71 @@ func deriveAddress(w *wallet, childnum uint32) (*address, error) {
 	}
 
 	// append childnum to derivation path and derive private key
-	path := append(w.DerivationPath, childnum)
+	path := append(w.DerivationPath, uint32(childnum))
 	key, err = derivePrivKey(key, path)
 	if err != nil {
 		return nil, err
 	}
 
-	// generate new address for derived key
-	addr, err := key.Address(net)
+	addr := ""
+	if w.Segwit {
+		pubkey, err := key.ECPubKey()
+		if err != nil {
+			return nil, err
+		}
+		keyHash := btcutil.Hash160(pubkey.SerializeCompressed())
+		scriptSig, err := txscript.NewScriptBuilder().AddOp(txscript.OP_0).AddData(keyHash).Script()
+		if err != nil {
+			return nil, err
+		}
+		rawAddress, err := btcutil.NewAddressScriptHash(scriptSig, net)
+		if err != nil {
+			return nil, err
+		}
+		addr = rawAddress.String()
+	} else {
+		// generate new address for derived key
+		rawAddress, err := key.Address(net)
+		if err != nil {
+			return nil, err
+		}
+		addr = rawAddress.String()
+	}
+
+	return &address{
+		Childnum:    childnum,
+		LastAddress: addr,
+	}, nil
+}
+
+func deriveSegWitAddress(w *wallet, childnum int) (*address, error) {
+	net, err := getNetworkFromString(w.Network)
+	if err != nil {
+		return nil, err
+	}
+
+	seed := seedFromMnemonic(w.Mnemonic)
+
+	// master key
+	key, err := hdkeychain.NewMaster(seed, net)
+	if err != nil {
+		return nil, err
+	}
+
+	// append childnum to derivation path and derive private key
+	path := append(w.DerivationPath, uint32(childnum))
+	key, err = derivePrivKey(key, path)
+	if err != nil {
+		return nil, err
+	}
+
+	pubkey, err := key.ECPubKey()
+	if err != nil {
+		return nil, err
+	}
+
+	witnessProgram := btcutil.Hash160(pubkey.SerializeCompressed())
+	addr, err := btcutil.NewAddressWitnessPubKeyHash(witnessProgram[:20], net)
```

Comment:
document number `20`  eg.  `scriptPubKey: HASH160 <20-byte-script-hash> EQUAL` and in other parts of code